### DISCUSSION TITLE
How to map enum.IntEnum to sqlalchemy.Enum?

### DISCUSSION THREAD

**Discussion by aaronsgithub** [sqlalchemy/sqlalchemy#10308] [Usage Questions]
Is it possible to use `enum.IntEnum` with `sa.Enum`?

```
import enum
import sqlalchemy as sa

class FileCategory(enum.IntEnum):
    DEFAULT = 0
    PROFILE_PHOTO = 1
    PORTFOLIO = 2
    
class File(Base):
    ...
    category: Mapped[FileCategory] = mapped_column(sa.Enum(FileCategory), default=lambda: FileCategory.DEFAULT)
```
When I try this, alembic stores the keys (as strings) instead of the values.

Note: I'm using PostgreSQL

**Comment by CaselIT** [comment]
Hi,

> Note: I'm using PostgreSQL

it's not clear if you want to use a postgresql enum or not in your use case.
If you want to use a pg enum, you need the names since pg uses names not numbers for the enum

**Reply by aaronsgithub** [reply]
I probably will end up using a PostgreSQL enum and drop the idea of using ints.

But for sake of completeness, how might I define a mapping from a python IntEnum to a PostgreSQL int column so that I can do `File(category=FileCategory.PORTFOLIO)` and it gets stored to 2 in the database.

**Reply by CaselIT** [reply]
I think just using a normal int / smallint as the column is sufficient. Since an int enum allows you do do `assert 0 == FileCategory.DEFAULT`

**Reply by iAnanich** [reply]
In general case, wouldn't it make sense to store as int when enum is IntEnum, unless DB support int enums?
Encountered similar question, but with sqlite, which doesn't support enums at all.
But I expect the simplicity of having enum object as model field. 
To me it seems like database should care about storing enum value - not it's name.

**Reply by iAnanich** [reply]
If going with storing in db as int, then field should be annotated as `Mapped[int]` or sqlalchemy will solve `Mapped[MyIntEnum]`?
(sorry, probably at this point I should open separate discussion...)

**Reply by CaselIT** [reply]
> Mapped[MyIntEnum]

This will store the enum names. You can tell sqlalchemy to store the values using https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.Enum.params.values_callable but I don't think it works of you pass integers there

**Reply by zzzeek** [reply]
it works if you send a values_callable that converts the ints to strings.

The enumerated datatype in SQLAlchemy is for strings.   an intenum would need to be a custom type.

**Reply by iAnanich** [reply]
So if `values_callable` returns `str(integer_value)` DB will save them as `'1'` or `'2'` strings, right? 
Well at least that's less space than whole names.
